datalad/support/gitrepo.py

```diff
@@ -671,21 +674,174 @@ def git_remote_update(self, name='', verbose=False):
+                raise ValueError("refspec specified without a remote. (%s)" %
+                                  refspec)
+            # No explicit remote to pull from.
+            # => get tracking branch:
+            tb = self.repo.active_branch.tracking_branch().name
+            if tb:
+                parts = tb.split('/')
+                assert len(parts) == 2
+                remote = self.repo.remote(parts[0])
+                refspec = parts[1]
+            else:
+                # No remote, no tracking branch
+                # => fail
+                raise ValueError("No remote specified to fetch from nor a "
+                                 "tracking branch is set up.")
+        else:
+            remote = self.repo.remote(remote)
+
+        fetch_url = \
+            remote.config_reader.get('fetchurl'
+                                     if remote.config_reader.has_option('fetchurl')
+                                     else 'url')
+        if fetch_url.startswith('ssh:'):
+            cnct = ssh_manager.get_connection(fetch_url)
+            cnct.open()
+            # TODO: with git <= 2.3 keep old mechanism:
+            #       with remote.repo.git.custom_environment(GIT_SSH="wrapper_script"):
+            with remote.repo.git.custom_environment(
+                    GIT_SSH_COMMAND="ssh -S %s" % cnct.ctrl_path):
+                remote.pull(refspec=refspec, progress=progress, **kwargs)
+                # TODO: progress +kwargs
+        else:
+            remote.pull(refspec=refspec, progress=progress, **kwargs)
+            # TODO: progress +kwargs
+
+    def push(self, remote=None, refspec=None, progress=None, all=False, **kwargs):
         """
         """
 
-        self._git_custom_command('', 'git fetch %s %s' % (options, name),
-                                 expect_stderr=True)
+        if remote is None:
+            if refspec is not None:
+                # conflicts with using tracking branch or fetch all remotes
+                # For now: Just fail.
+                # TODO: May be check whether it fits to tracking branch
+                raise ValueError("refspec specified without a remote. (%s)" %
+                                  refspec)
+            if all:
+                remotes_to_push = self.repo.remotes
+            else:
+                # No explicit remote to fetch.
+                # => get tracking branch:
+                tb = self.repo.active_branch.tracking_branch().name
+                if tb:
+                    parts = tb.split('/')
+                    assert len(parts) == 2
+                    remotes_to_push = [self.repo.remote(parts[0])]
+                    refspec = parts[1]
+                else:
+                    # No remote, no tracking branch
+                    # => fail
+                    raise ValueError("No remote specified to fetch from nor a "
+                                     "tracking branch is set up.")
```

Comment:
so indeed a 3rd copy and this one is not even covered by tests at all... centralization indeed would help at least to cover the common code